### PR TITLE
#226 - review and fix spotbug findings

### LINKED ISSUE #226: Upgrade findbugs to spotbugs
* the current findbugs plugin does not work with java 8
* the plugin is not being executed in the CI

Therefore:
* upgrade to spotbugs
* ci build with spotbugs after the main run

### PR DESCRIPTION
**Pull request by sparsick**
fixed #226

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -236,6 +236,18 @@
...
+        <dependency>
+            <groupId>org.codehaus.groovy</groupId>
+            <artifactId>groovy</artifactId>
+            <version>${groovy.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>com.github.spotbugs</groupId>
+            <artifactId>spotbugs-annotations</artifactId>
+            <version>4.2.1</version>
+        </dependency>
```
Comment (bmarwell) [inline pom.xml:248]: provided scope is sufficient

File: src/main/groovy/org/reficio/p2/resolver/eclipse/impl/DefaultEclipseResolver.groovy [hunk 2]
```diff
@@ -18,12 +18,12 @@
  */
 package org.reficio.p2.resolver.eclipse.impl
 
-import org.apache.commons.io.FilenameUtils
 import org.reficio.p2.logger.Logger
 import org.reficio.p2.resolver.eclipse.EclipseResolutionRequest
 import org.reficio.p2.resolver.eclipse.EclipseResolutionResponse
 import org.reficio.p2.resolver.eclipse.EclipseResolver
 
+@edu.umd.cs.findbugs.annotations.SuppressFBWarnings("SE_NO_SERIALVERSIONID")
```
Comment (bmarwell) [inline src/main/groovy/org/reficio/p2/resolver/eclipse/impl/DefaultEclipseResolver.groovy:26]: No import? You use imports in other classes.

File: src/main/java/org/reficio/p2/P2Validator.java [hunk 3]
```diff
@@ -60,19 +60,19 @@ public static void validateArtifactConfig(P2Artifact p2Artifact, ResolvedArtifac
...
-                        "\n\tWATCH OUT! Setting <override> to true will re-bundle the artifact!\n", resolvedArtifact.getArtifact().toString());
+                        "%n%n\tJar [%s] is already a bundle. " +
+                        "%n\tBND instructions are specified, but the <override> flag is set to false." +
+                        "%n\tEither remove the instructions or set the <override> flag to true." +
+                        "%n\tWATCH OUT! Setting <override> to true will re-bundle the artifact!%n", resolvedArtifact.getArtifact().toString());
                 throw new RuntimeException(message);
             }
             // artifact is a bundle and somebody specified singleton flag without override
             if (!p2Artifact.shouldOverrideManifest() && p2Artifact.isSingleton()) {
                 String message = String.format("p2-maven-plugin misconfiguration" +
```
Comment (bmarwell) [inline src/main/java/org/reficio/p2/P2Validator.java:71]: While at it, you should also explicitly define a locale (`String.format(Locale.ENGLISH, "…"`).

File: src/main/java/org/reficio/p2/P2Mojo.java [hunk 4]
```diff
@@ -545,14 +548,18 @@ private void executeCategoryPublisher() throws AbstractMojoExecutionException, I
...
-            InputStream is = getClass().getResourceAsStream(DEFAULT_CATEGORY_CLASSPATH_LOCATION + DEFAULT_CATEGORY_FILE);
-            File destinationFolder = new File(destinationDirectory);
-            destinationFolder.mkdirs();
-            File categoryDefinitionFile = new File(destinationFolder, DEFAULT_CATEGORY_FILE);
-            FileWriter writer = new FileWriter(categoryDefinitionFile);
-            IOUtils.copy(is, writer, "UTF-8");
+            File categoryDefinitionFile;
+            FileWriter writer;
+            try (InputStream is = getClass().getResourceAsStream(DEFAULT_CATEGORY_CLASSPATH_LOCATION + DEFAULT_CATEGORY_FILE)) {
+                File destinationFolder = new File(destinationDirectory);
```
Comment (bmarwell) [inline src/main/java/org/reficio/p2/P2Mojo.java:557]: Any reason why we do not want to define an encoding explicitly? System.getDefaultEncoding() would also work for me.

**Review by bmarwell — CHANGES_REQUESTED** [review]

File: src/main/groovy/org/reficio/p2/resolver/eclipse/impl/DefaultEclipseResolver.groovy (hunk 2 above)
Comment (sparsick) [reply src/main/groovy/org/reficio/p2/resolver/eclipse/impl/DefaultEclipseResolver.groovy:26]: I think my idea fools me here. I will add it manually. Thanks

File: src/main/java/org/reficio/p2/P2Mojo.java (hunk 4 above)
Comment (sparsick) [reply src/main/java/org/reficio/p2/P2Mojo.java:557]: I dont know. It is the original implementation. I will run the ITs again with default encoding and we will see what happens

File: src/main/java/org/reficio/p2/P2Validator.java (hunk 3 above)
Comment (sparsick) [reply src/main/java/org/reficio/p2/P2Validator.java:71]: Will do. Thanks

File: src/main/java/org/reficio/p2/P2Mojo.java (hunk 4 above)
Comment (bmarwell) [reply src/main/java/org/reficio/p2/P2Mojo.java:557]: I see that IOUtils uses UTF-8 explicitly. I think you should just use UTF-8 then.

File: src/main/groovy/org/reficio/p2/resolver/maven/impl/facade/AetherSonatypeFacade.groovy [hunk 5]
```diff
@@ -37,6 +39,7 @@ import org.sonatype.aether.util.graph.PreorderNodeListGenerator
  *         http://www.reficio.org
  * @since 1.1.0
  */
+@edu.umd.cs.findbugs.annotations.SuppressFBWarnings("SE_NO_SERIALVERSIONID")
```
Comment (bmarwell) [inline src/main/groovy/org/reficio/p2/resolver/maven/impl/facade/AetherSonatypeFacade.groovy:42]: Now you have both an import AND a full qualified class.

File: src/main/java/org/reficio/p2/FeatureBuilder.java [hunk 6]
```diff
@@ -56,6 +56,7 @@ public FeatureBuilder(P2FeatureDefinition p2FeatureDefintion,
 	//cache this so that the same timestamp is used
 	private String featureTimeStamp;
 	
+	@edu.umd.cs.findbugs.annotations.SuppressFBWarnings("RV_RETURN_VALUE_IGNORED_BAD_PRACTICE")
```
Comment (bmarwell) [inline src/main/java/org/reficio/p2/FeatureBuilder.java:59]: more of qualified imports

File: src/main/java/org/reficio/p2/P2Mojo.java [hunk 7]
```diff
@@ -80,6 +84,7 @@
         requiresDependencyResolution = ResolutionScope.RUNTIME,
         requiresDependencyCollection = ResolutionScope.RUNTIME
 )
+@edu.umd.cs.findbugs.annotations.SuppressFBWarnings("RV_RETURN_VALUE_IGNORED_BAD_PRACTICE")
```
Comment (bmarwell) [inline src/main/java/org/reficio/p2/P2Mojo.java:87]: more of qualified imports

File: src/main/java/org/reficio/p2/utils/JarUtils.java [hunk 8]
```diff
@@ -78,6 +78,7 @@ public static void adjustSnapshotOutputVersion(File inputFile, File outputFile,
      * @param timestamp - timestamp
      *
      */
+    @edu.umd.cs.findbugs.annotations.SuppressFBWarnings("RV_RETURN_VALUE_IGNORED_BAD_PRACTICE")
```
Comment (bmarwell) [inline src/main/java/org/reficio/p2/utils/JarUtils.java:81]: more of qualified imports

File: src/main/groovy/org/reficio/p2/resolver/eclipse/impl/FileBinaryCategory.groovy [hunk 9]
```diff
@@ -18,6 +18,9 @@
  */
 package org.reficio.p2.resolver.eclipse.impl
 
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+@edu.umd.cs.findbugs.annotations.SuppressFBWarnings("SE_NO_SERIALVERSIONID")
```
Comment (bmarwell) [inline src/main/groovy/org/reficio/p2/resolver/eclipse/impl/FileBinaryCategory.groovy:23]: Now you have an import and don’t use it ;)

File: src/main/groovy/org/reficio/p2/resolver/eclipse/impl/DefaultEclipseResolver.groovy [hunk 10]
```diff
@@ -18,12 +18,14 @@
...
 
-import org.apache.commons.io.FilenameUtils
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
 import org.reficio.p2.logger.Logger
 import org.reficio.p2.resolver.eclipse.EclipseResolutionRequest
 import org.reficio.p2.resolver.eclipse.EclipseResolutionResponse
 import org.reficio.p2.resolver.eclipse.EclipseResolver
 
+@edu.umd.cs.findbugs.annotations.SuppressFBWarnings("SE_NO_SERIALVERSIONID")
```
Comment (bmarwell) [inline src/main/groovy/org/reficio/p2/resolver/eclipse/impl/DefaultEclipseResolver.groovy:28]: Now you have an import and don’t use it ;)

File: src/main/groovy/org/reficio/p2/resolver/maven/impl/AetherResolver.groovy [hunk 11]
```diff
@@ -32,6 +34,7 @@ import org.reficio.p2.resolver.maven.ResolvedArtifact
  *         http://www.reficio.org
  * @since 1.0.0
  */
+@edu.umd.cs.findbugs.annotations.SuppressFBWarnings("SE_NO_SERIALVERSIONID")
```
Comment (bmarwell) [inline src/main/groovy/org/reficio/p2/resolver/maven/impl/AetherResolver.groovy:37]: full qualified

File: src/main/groovy/org/reficio/p2/resolver/maven/impl/facade/AetherEclipseFacade.groovy [hunk 12]
```diff
@@ -37,6 +39,7 @@ import org.reficio.p2.resolver.maven.Artifact
  *         http://www.reficio.org
  * @since 1.1.0
  */
+@edu.umd.cs.findbugs.annotations.SuppressFBWarnings("SE_NO_SERIALVERSIONID")
```
Comment (bmarwell) [inline src/main/groovy/org/reficio/p2/resolver/maven/impl/facade/AetherEclipseFacade.groovy:42]: full qualified

File: src/main/groovy/org/reficio/p2/resolver/maven/impl/facade/AetherFacade.groovy [hunk 13]
```diff
@@ -18,6 +18,8 @@
  */
 package org.reficio.p2.resolver.maven.impl.facade
 
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
 import org.reficio.p2.resolver.maven.Artifact
```
Comment (bmarwell) [inline src/main/groovy/org/reficio/p2/resolver/maven/impl/facade/AetherFacade.groovy:23]: unused

**Review by bmarwell — CHANGES_REQUESTED** [review]
some small oversights/nits

File: src/main/groovy/org/reficio/p2/resolver/maven/impl/facade/AetherSonatypeFacade.groovy (hunk 5 above)
Comment (sparsick) [reply src/main/groovy/org/reficio/p2/resolver/maven/impl/facade/AetherSonatypeFacade.groovy]: argggh yes thanks :)

File: src/main/groovy/org/reficio/p2/resolver/eclipse/impl/FileBinaryCategory.groovy (hunk 9 above)
Comment (sparsick) [reply src/main/groovy/org/reficio/p2/resolver/eclipse/impl/FileBinaryCategory.groovy]: hopefully fixed

File: src/main/groovy/org/reficio/p2/resolver/eclipse/impl/DefaultEclipseResolver.groovy [hunk 14]
```diff
@@ -18,12 +18,14 @@
...
 
-import org.apache.commons.io.FilenameUtils
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
+
 import org.reficio.p2.logger.Logger
 import org.reficio.p2.resolver.eclipse.EclipseResolutionRequest
 import org.reficio.p2.resolver.eclipse.EclipseResolutionResponse
 import org.reficio.p2.resolver.eclipse.EclipseResolver
 
+@edu.umd.cs.findbugs.annotations.SuppressFBWarnings("SE_NO_SERIALVERSIONID")
```
Comment (bmarwell) [inline src/main/groovy/org/reficio/p2/resolver/eclipse/impl/DefaultEclipseResolver.groovy:28]: Nope, still there…

File: src/main/java/org/reficio/p2/P2Mojo.java [hunk 15]
```diff
@@ -545,14 +549,18 @@ private void executeCategoryPublisher() throws AbstractMojoExecutionException, I
         publisher.execute();
     }
 
+    @SuppressFBWarnings("DM_DEFAULT_ENCODING")
```
Comment (bmarwell) [inline src/main/java/org/reficio/p2/P2Mojo.java:552]: Remove this warning and…

File: src/main/java/org/reficio/p2/P2Mojo.java [hunk 16]
```diff
@@ -545,14 +549,18 @@ private void executeCategoryPublisher() throws AbstractMojoExecutionException, I
...
-            File categoryDefinitionFile = new File(destinationFolder, DEFAULT_CATEGORY_FILE);
-            FileWriter writer = new FileWriter(categoryDefinitionFile);
-            IOUtils.copy(is, writer, "UTF-8");
+            File categoryDefinitionFile;
+            FileWriter writer;
+            try (InputStream is = getClass().getResourceAsStream(DEFAULT_CATEGORY_CLASSPATH_LOCATION + DEFAULT_CATEGORY_FILE)) {
+                File destinationFolder = new File(destinationDirectory);
+                destinationFolder.mkdirs();
+                categoryDefinitionFile = new File(destinationFolder, DEFAULT_CATEGORY_FILE);
+                writer = new FileWriter(categoryDefinitionFile);
```
Comment (bmarwell) [inline src/main/java/org/reficio/p2/P2Mojo.java:561]: … use UTF-8 here, because the next line uses UTF-8 anyway.

**Review by bmarwell — CHANGES_REQUESTED** [review]
Still some comments not picked up, sorry =)

File: src/main/java/org/reficio/p2/P2Mojo.java (hunk 16 above)
Comment (bmarwell) [reply src/main/java/org/reficio/p2/P2Mojo.java:561]: … this comment should have highlighted the `new FileWriter()` call.

Comment (sparsick) [reply src/main/java/org/reficio/p2/P2Mojo.java:561]: > … this comment should have highlighted the `new FileWriter()` call.

Yes that true. But Filewriter constructor has no constructor with encoding. so I have to replace it by `new OutputStreamWriter(new FileOutputStream(categoryDefinitionFile), StandardCharsets.UTF_8);

**Review by bmarwell — APPROVED** [review]

**Pull request merged**